### PR TITLE
Add letter/word/line spacing factors to metadata

### PR DESCRIPTION
**Pull request by ubruhin**
These three factors are now part of the [font] metadata section and can
be used by a font layout engine to calculate the spacing between
letters, words and lines:

- letter_spacing
- word_spacing
- line_spacing

See #14

Note: Of course it would be nice to add some graphics to illustrate how these spacing values are used exactly. But that requires some effort which I don't have the time for, I think this can be done some day later ;)

### CODE REVIEW

File: SPECIFICATION.md [hunk 1]
```diff
@@ -186,6 +189,9 @@ TODO
 | version | The version of this font. SHOULD follow semantic versioning. | 1 | `0.4.1` |
 | author | The name of the copyright owner, in the format `Name <email>`. The email part is optional. | 0-n | `Max Müller <max@foo>` |
 | license | The SPDX identifier for the license of this font. Create multiple `License` entries if the font is published under multiple licenses. | 1-n | `Apache-2.0` |
+| letter_spacing | Horizontal space which the layout engine should place between glyphs. Same unit as for polylines. | 1 | `3` (= 1/3 of font height) |
+| word_spacing | Horizontal space which the layout engine should place between words. Same unit as for polylines. | 1 | `6.75` (= 3/4 of font height) |
+| line_spacing | Vertical distance which the layout engine should use for multi-line texts. Same unit as for polylines. | 1 | `15` (= 5/3 of font height) |
```
Comment (dbrgn) [inline SPECIFICATION.md]: I'd make all those values optional, with sensible defaults.

**Review by dbrgn — COMMENTED** [review]
This refs #5.

I'm still not really convinced if a font should be responsible for things like line spacing. At least not in absolute values. Usually the font engine chooses values that seem sensible to the application.

What if instead of absolute values we would provide multipliers, which are 1 by default? A font that does not require any special handling would not set them (or set them to 1). A condensed font would set a letter spacing value of 0.8. And a font with a larger than usual descender height would set line spacing to 1.5.

RFC @rnestler

**Comment by ubruhin** [comment]
> I'm still not really convinced if a font should be responsible for things like line spacing.

For a simple renderer it's really hard to decide what line spacing should be used. So it would be *very* helpful to have a suggested line spacing defined in the font file. Also the font designer probably knows better than the renderer with what line spacing the font looks best.

But of course, the values defined in the font files are just a suggestion for the renderer. The renderer can decide on its own if and how these values are used for rendering.

> At least not in absolute values.

Actually the values aren't absolute. Just divide them by 9 to get a multiplier ;) So a line spacing of 15 actually means 15/9 = 1.67 times the font height.

> What if instead of absolute values we would provide multipliers, which are 1 by default?

This is actually exactly what LFF does. But I don't like that because then the default line spacing of 1.67 is hardcoded in the application's source code. It took me quite a lot of time to understand how the line spacing in LibreCAD is calculated. Having the *real* factor (either 1.67 or 15, doesn't matter) written in the font file would make it much easier for every user to understand how line spacing is calculated.

Of course the renderer can add an additional factor to the line spacing to decrease/increase line spacing if needed.

See also my explanation in https://github.com/fontobene/fontobene/issues/14#issuecomment-370072154.

**Comment by ubruhin** [comment]
Rebased.

**Comment by dbrgn** [comment]
> Actually the values aren't absolute. Just divide them by 9 to get a multiplier ;) So a line spacing of 15 actually means 15/9 = 1.67 times the font height.

No, what I meant is that the values should be relative to the defaults that the renderer chooses, not relative to the font size.

I'm not against proposing default values in the specification though.

**Comment by ubruhin** [comment]
> No, what I meant is that the values should be relative to the defaults that the renderer chooses, not relative to the font size.

I see. But then the question is still: what defaults should the renderer choose? How should the renderer know what line spacing is suitable for a given font?

> I'm not against proposing default values in the specification though.

You mean that in the specification we would write something like "default line spacing is 1.67 times the font height"? What would be the advantage in contrast to specifying the line spacing in every font file?

I see that it would be nice to make these values optional in the font files. But then the default values need to be defined in the fontobene specifications (if not, every renderer would so something different). That would be kind of "magic number" in the specifications which I don't like. My opinion is that the font designer knows best which line spacing is suitable for his font, so why shouldn't he specify that line spacing right in the font file?

**Comment by dbrgn** [comment]
> I see. But then the question is still: what defaults should the renderer choose?

The renderer should probably choose the values depending on the application. If you're labelling PCBs, then it's much more important that the text is clearly readable, so you'd probably choose larger letter and word spacing. If on the other hand you're rendering text in a context where you want to save space, you'd probably choose smaller defaults.

> How should the renderer know what line spacing is suitable for a given font?

It can choose good defaults for most fonts. If a font has special requirements (e.g. if it's a condensed font), then the font can choose a scaling factor to correct for that.

> You mean that in the specification we would write something like "default line spacing is 1.67 times the font height"?

I thought about something like "we found that using a default spacing of 1.67 times the font height produces good results in most cases".

> What would be the advantage in contrast to specifying the line spacing in every font file?

The font designer does not know the context in which the font is used.

> I see that it would be nice to make these values optional in the font files. But then the default values need to be defined in the fontobene specifications (if not, every renderer would so something different).

It's quite simple: If these values are multipliers, then they all default to 1.

> My opinion is that the font designer knows best which line spacing is suitable for his font

I think the font designer should tweak the parameters to suit her font (e.g. by saying "this font needs larger line spacing because the 'g' glyph reaches down lower than other fonts"), but the implementer  decides what base spacing to use.

@rnestler RFC :)

**Comment by ubruhin** [comment]
> @rnestler RFC :)

@rnestler would probably say (as in https://github.com/fontobene/fontobene-converters/issues/1#issuecomment-375237007):

This argument also works the other way around: The font designer can choose what line spacing is suitable for his font (e.g. 1.67). The application can apply an additional factor on the spacing, depending on the needs of the application. If high readability is required, it could use factor 1.2, if space saving is required, it could choose 0.8.

SCNR :grin:

**Comment by ubruhin** [comment]
Another advantage of specifying the whole line spacing (e.g. 1.67) in font files:

Font designer require a font viewer to see how the font looks like. So they use such viewers to review and adjust their fonts. But if the line spacing is up to the application, one viewer may use 1.5 as line spacing, and another viewer uses 1.8. When using the viewer with 1.8, the font designer sees that the line spacing is too high for his font and sets the factor (which defaults to 1.0) in his font file to 0.8 to make the font looking nice.

Then you open the same font file with the other viewer (with 1.5) and the font is completely messed up because of too less line spacing.

When specifying the whole line spacing in the font file, the result is independent of the viewer, because every viewer will probably just use the line spacing specified in the font file without any additional factor (because there's no reason to add another factor).

**Comment by dbrgn** [comment]
> This argument also works the other way around: The font designer can choose what line spacing is suitable for his font (e.g. 1.67). The application can apply an additional factor on the spacing, depending on the needs of the application.

But the font designer designs for one font. The application must work with many fonts, and they should have consistent look and feel (unless the font has special requirements). Thus the application should choose the defaults, and the fonts that have special requirements can deviate from that.

Your argument with the font viewer is solved by providing a recommendation in the specification.

Maybe we can look at how established formats like OpenType handle this? As far as I know the font does not specify a fixed line spacing, but instead line spacing and line height are handled by the viewer.

**Comment by dbrgn** [comment]
By the way, the values are not well-defined in the current draft.

- I think it's common to define line spacing by choosing the line height (baseline to baseline). What is being used for fontobene?
- Word spacing is ambiguous. When will word spacing be applied? Between two words? For every whitespace character? What if two spaces are inserted? What about tab characters and [thin spaces](http://jkorpela.fi/chars/spaces.html)? Should the whitespace characters be specified in the font, or not? And if yes, how do you declare them, since there is no polyline?

**Comment by rnestler** [comment]
I would standardize the line-spacing, something like https://en.wikipedia.org/wiki/Leading. With a given font height the line spacing should be the same even if I change fonts.

**Comment by dbrgn** [comment]
@ubruhin and I discussed this a bit.

We decided to add "whitespace polylines" to the specification. That allows to specify things like spaces, wide spaces, etc.

With this, the need for word spacing is gone. Left are letter spacing and line spacing.

Character spacing should be specified in terms relative to the font size. Otherwise we'll get strange effects in combination with potential kerning values.

Even though I defended the use of multipliers for specifying line spacing, I think in this case consistency is important and I'd vote for specifying both line- and character spacing in terms of our unit system.

Another result of our discussion: There should be sane defaults for these two metrics that are listed in the specification. The values themselves are optional inside the font file and should only be specified when deviating from the default values, because the font has "special needs". An implementation is of course free to change the values, e.g. by multiplying them by a factor.

@ubruhin is this correct? If yes, can you update the PR?

**Comment by ubruhin** [comment]
> With a given font height the line spacing should be the same even if I change fonts.

Of course that would be nice, but there can be fonts which have smaller or larger ascender/descender and thus need to adjust the line spacing accordingly to still look nice. So IMHO it makes sense to specify that in the font file. If in a specific use case it's more important to have the same line spacing for all fonts, the application is free to ignore the value from the font file and use a fixed value instead.

> @ubruhin is this correct? If yes, can you update the PR?

Yep, I'll work on specs for the things you wrote.

**Comment by ubruhin** [comment]
Letter/word spacing is now in a separate PR: #32

For line spacing I will also open a new PR.

So I'll close this one as we won't merge it.

**Pull request closed without merging**